include/seastar/coroutine/async_generator.hh

```diff
@@ -0,0 +1,461 @@
+// introduced std::generator, which provides a synchronous coroutine
+// mechanism for generating ranges. in contrast, async_generator offers
+// asynchronous generation of element sequences.
+namespace seastar::coroutine::experimental {
+
+template<typename Ref, typename Value = void>
+class async_generator;
+
+namespace internal {
+
+template <typename Yielded> class next_awaiter;
+
+template <typename Yielded>
+class async_generator_promise_base : public seastar::task {
+protected:
+    // a glvalue yield expression is passed to co_yield as its operand. and
+    // the object denoted by this expression is guaranteed to live until the
+    // coroutine resumes. we take advantage of this fact by storing only a
+    // pointer to the denoted object in the promise as long as the result of
+    // dereferencing this pointer is convertible to the Ref type.
+    std::add_pointer_t<Yielded> _value = nullptr;
+
+protected:
+    std::exception_ptr _exception;
+    std::coroutine_handle<> _consumer;
+    task* _waiting_task = nullptr;
+
+    /// awaiter returned by the generator when it produces a new element
+    ///
+    /// There are different combinations of expression types passed to
+    /// \c co_yield and \c Ref. In most cases, zero copies are made. Copies
+    /// are only necessary when \c co_yield requires type conversion.
+    ///
+    /// The following table summarizes the number of copies made for different
+    /// scenarios:
+    ///
+    /// | Ref       | co_yield const T& | co_yield T& | co_yield T&& | co_yield U&& |
+    /// | --------- | ----------------- | ----------- | ------------ | ------------ |
+    /// | T         | 0                 | 0           | 0            | 1            |
+    /// | const T&  | 0                 | 0           | 0            | 1            |
+    /// | T&        | ill-formed        | 0           | ill-formed   | ill-formed   |
+    /// | T&&       | ill-formed        | ill-formed  | 0            | 1            |
+    /// | const T&& | ill-formed        | ill-formed  | 0            | 1            |
+    ///
+    /// When no copies are required, \c yield_awaiter is used. Otherwise,
+    /// \c copy_awaiter is used. The latter converts \c U to \c T, and keeps the converted
+    /// value in it.
+    class yield_awaiter final {
+        async_generator_promise_base* _promise;
+        std::coroutine_handle<> _consumer;
+    public:
+        yield_awaiter(async_generator_promise_base* promise,
+                      std::coroutine_handle<> consumer) noexcept
+            : _promise{promise}
+            , _consumer{consumer}
+        {}
+        bool await_ready() const noexcept {
+            return false;
+        }
+        template <typename Promise>
+        std::coroutine_handle<> await_suspend(std::coroutine_handle<Promise> producer) noexcept {
+            _promise->_waiting_task = &producer.promise();
+            return _consumer;
```

Comment:
Here we have symmetric transfer, which is nice, the consumer is awakened without going through the scheduler. However, we must make at least one of the two directions has a preemption check.

I think it's okay now because await_ready always returns false, but I also think await_ready shouldn't return false.

If we are able to do symmetric transfer in both directions, then the ping-pong problem is much reduced: the two coroutines exchange control until one of them detects that preemption is needed or has to block itself.

Reply:
> However, we must make at least one of the two directions has a preemption check.

i will take a closer look to see how to season the awaiter gently.

Reply:
should be fixed in the latest revision, which yields on the producer's end.